### ISSUE TITLE
Default concurrency causing issues.

### ISSUE THREAD

**Issue by amardomingo** [roboll/helmfile#484]
Hi!

I don't know if this is related to #411 , but we are having some issues that seems to be related to the concurrency flag

When running without specifying `--concurrency` in `sync` or `apply` , we are seeing a bunch of errors like this:
```
Upgrading <path-to-chart-1>
UPGRADE FAILED
ROLLING BACK
Error: release: "<name-of-chart-2>.<some-version>" already exists
Error: UPGRADE FAILED: release: "<name-of-chart-2>" already exists
```

As far as I can tell, it looks like it is mixing up charts and values from different entries in our helmfile (we have about 9 releases in this particular helmfile). Going with any value other than concurrency 1 seems to give the errors.

**Comment by sstarcher** [comment]
@amardomingo are you defining the same release multiple times?  I have somewhere between 2-3 dozen releases and I have not seen any concurrency issues except for my box slowing to a crawl.  I typically run it with 8 concurrency.

**Comment by amardomingo** [comment]
@sstarcher unless I am absolutely blind, no. I have a bunch of releases with similar names (for example, I have "istio", "istio-init", "cassandra-istio" and so on), but I don't have any duplicated releases.

Anyhow, I have changed the binary of helmfile from direct downloading it from the releases page to installing it with homebrew, and I don't seem to be able to replicate the problem again. I know that does not make much sense (if any at all), but it's the only thing I've changed about helmfile lately.

Should I close  the issue?

**Comment by mumoshu** [comment]
@amardomingo Hey! Thanks for the report.

You shouldn't close this! As myself being a helmfile user, I'll report here if I get to reproduce this. Hard to say reinstalling or changing the source of helmfile binary makes difference, but worth noting as we have no idea why this is happening yet.

**Comment by andytom** [comment]
This sounds similar to an issue we had when running diff with concurrency set to 3 (https://github.com/roboll/helmfile/issues/410), after updating everything this just seemed to stop so I closed the issue but I never found the root cause, I'm just mentioning this since it might help with the debugging.

**Comment by mumoshu** [comment]
@amardomingo Btw, which version of helmfile are you using?

**Comment by amardomingo** [comment]
Currently I'm using v0.45.3 in Mac, installed with homebrew. 

When I created the issue I was using a direct download from the releases page, I think it migth have been v0.45.2, but when I ran "helmfile -v" it only said "version". I currently can't replicate that behaviour, so I'm starting to think that there was something wrong with my system.

I just downloaded all v0.45.x versions from he releases page in a colleague's Mac and he has not been able to replicate that behaviour either :-(

**Comment by mumoshu** [comment]
@amardomingo Thanks for confirming! I guess it's fixed in recent versions of helmfile then.

I've heard about possible concurrency-related bugs like yours before. I wasn't aware of the exact issue, but hoping to fix it altogether, I've refactored all the duplicated code/logics around the concurrency feature as a part of #442, which has released as part of [v0.43.0](https://github.com/roboll/helmfile/releases/tag/v0.43.0). So probably helmfile, starting v0.43.0, won't see the issue.

Regarding your `helmfile -v` output, perhaps it may be due to that it was installed via `go get`. We have a dedicated issue for #304, but have no solution at the moment (yet).

**Comment by mumoshu** [comment]
Closing this as already fixed. Please feel free to reopen if you still encounter this using recent versions of helmfile!
